### PR TITLE
Added benefits of OOP in JS - #1269 resolved

### LINKED ISSUE #1269: Add "Why OOP in JS?" section to JS 7 - OOP
In the [JS 7 OOP lesson](https://github.com/Techtonica/curriculum/blob/master/javascript/javascript-7-oop.md) it doesn't really cover why one should learn OOP in JS, considering that JS code and libraries tend not to use OOP and instead skew more functional.

Look at the JavaScript section in the readme to be aware of what gets covered before this lesson.

Let's add a section immediately following "Motivation" called "Why learn OOP in JavaScript?". It should cover the following points:
- [ ] While not as popular in JS as some languages, React uses classes for its components (find link showing this)
- [ ] While its OOP style is unique, it shouldn't be considered second-rate and can be used to build good OOP code
- [ ] Participants will need to understand OOP for technical interviews and to learn other languages so it's good to learn it now
- [ ] Please find any blogs or videos explaining the benefits of OOP in JS

## Ensure inclusion of 
- dependency inversion
- Liskov substitution
- [SOLID](https://en.wikipedia.org/wiki/SOLID)

### PR DESCRIPTION
**Pull request by Rukmini-Meda**
Fixes #1269 

Changes proposed:
- Added benefits of OOP, a brief introduction
- Included resources for various concepts

Concepts covered:
- Introduction to functional programming
- Importance of OOP
- SOLID principles
- JS frameworks

@alodahl @Kundan28 Please review and merge this PR. Thank you.

### CODE REVIEW

File: javascript/javascript-7-oop.md [hunk 1]
```diff
@@ -27,6 +27,29 @@
 
 Virtually all companies will use object-oriented programming in their code. For example, when you do a Yelp search for "burgers", each area on the screen that describes each recommended restaurant is an instance of a class. The information is probably abstracted into a class of "Restaurant", of which "Marlowe" and "Super Duper Burgers" are instances of that class.
 
+### Benefits of OOP in JS
+
+- Javascript is mostly used a functional programming language more than an object oriented language like C++, Java. Functional programming in JS can be elegantly used to code and organize the same. The following is a resource to understand the advantages of functional programming in JS.
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:32]:
```suggestion
- JavaScript is mostly used as a functional programming language more than an object oriented language like C++ or Java. Functional programming in JS can be elegantly used to organize code. Learn more about the advantages of [Functional Programming in JS](https://blog.bitsrc.io/functional-programming-in-javascript-how-and-why-94e7a97343b).
```

File: javascript/javascript-7-oop.md [hunk 2]
```diff
@@ -27,6 +27,29 @@
...
+
+- [Functional Programming in JS](https://blog.bitsrc.io/functional-programming-in-javascript-how-and-why-94e7a97343b)
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:34]:
```suggestion
```

File: javascript/javascript-7-oop.md [hunk 3]
```diff
@@ -27,6 +27,29 @@
...
+
+- Javascript has a unique OOP style athough its less famous. It can also be used to produce reliable code. Although Javascript hasnt got famous OOP concepts, the frameworks like React, Angular and Vue rely a lot on the same and have got increased popularity amongst developers world wide in many projects.
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:36]:
```suggestion
- JavaScript has a unique, reliable OOP style, although it is less famous.  JS frameworks like [React](https://stackoverflow.com/questions/35764800/reactjs-object-oriented), Angular and Vue rely on OOP and have recently increased in popularity amongst developers world wide.
```

File: javascript/javascript-7-oop.md [hunk 4]
```diff
@@ -27,6 +27,29 @@
...
+
+- [Object oriented programming in React](https://stackoverflow.com/questions/35764800/reactjs-object-oriented)
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:38]:
```suggestion
```

File: javascript/javascript-7-oop.md [hunk 5]
```diff
@@ -27,6 +27,29 @@
...
+
+- Javascript and its frameworks have become must skills for any web developer to master. It's important for students to know for technical interviews as well as to learn other frameworks or languages which have a few ideas based on them.
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:40]:
```suggestion
- JavaScript and its frameworks have become must-have skills for any web developer to master. It's important for students to know for technical interviews as well as to learn other frameworks or languages which have a few ideas based on them. 
```
Can you rephrase this sentence?  I don't understand it.  "It's important for students to know for technical interviews as well as to learn other frameworks or languages which have a few ideas based on them. "

File: javascript/javascript-7-oop.md [hunk 6]
```diff
@@ -27,6 +27,29 @@
...
+
+- [Functional Programming in JS](https://blog.bitsrc.io/functional-programming-in-javascript-how-and-why-94e7a97343b)
+
+- Javascript has a unique OOP style athough its less famous. It can also be used to produce reliable code. Although Javascript hasnt got famous OOP concepts, the frameworks like React, Angular and Vue rely a lot on the same and have got increased popularity amongst developers world wide in many projects.
+
+- [Object oriented programming in React](https://stackoverflow.com/questions/35764800/reactjs-object-oriented)
+
+- Javascript and its frameworks have become must skills for any web developer to master. It's important for students to know for technical interviews as well as to learn other frameworks or languages which have a few ideas based on them. 
+
+- Software development lifecycle has to go through many different stages. Writing clean, organized, reusable code is atmost necessry for developers to maintain projects. Hence, the five principle of SOLID applied in Javascript are important to master the ability to write clean and professional code. The following are a few resources to understand SOLID principles, the importance of OOP in JS.
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:42]:
```suggestion
- The software development lifecycle has to go through many different stages. Writing clean, organized, reusable code is essential for developers to maintain projects. Hence, the five principles of SOLID applied in JavaScript are important to master the ability to write clean and professional code. Here are a few resources on SOLID principles and the importance of OOP in JS:
```

File: javascript/javascript-7-oop.md [hunk 7]
```diff
@@ -27,6 +27,29 @@
...
+
+
+- [SOLID in JS 1](https://medium.com/@cramirez92/s-o-l-i-d-the-first-5-priciples-of-object-oriented-design-with-javascript-790f6ac9b9fa)
+
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:46]:
```suggestion
```

File: javascript/javascript-7-oop.md [hunk 8]
```diff
@@ -27,6 +27,29 @@
...
+- [SOLID in JS 2](https://dev.to/francescoxx/solid-principles-in-javascript-3pek)
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:47]: Can you re-title these?  This makes it look like "JS1" and "JS2" are official things, when really you are just listing them.  I'd suggest using the articles' titles here.

File: javascript/javascript-7-oop.md [hunk 9]
```diff
@@ -27,6 +27,29 @@
...
+
+### Materials
+
+- [Video - Why OOP was created and advantages](https://www.youtube.com/watch?v=AjR2Rc9wQ6s)
```
Comment (alodahl) [inline javascript/javascript-7-oop.md:51]:
```suggestion
- [Video (4 min) - Why was OOP Invented?](https://www.youtube.com/watch?v=AjR2Rc9wQ6s)
```

**Comment by alodahl** [comment]
Oh also, were you able to include these from the issue?
"dependency inversion
Liskov substitution"

**Comment by Rukmini-Meda** [comment]
> Oh also, were you able to include these from the issue?
> "dependency inversion
> Liskov substitution"

Yes, they are covered in SOLID - they are the 'L' and 'D' part in SOLID.

**Comment by Rukmini-Meda** [comment]
@alodahl I have made the changes you suggested and also updated the projected time. If there are no more changes, please merge. Thank you.

**Comment by alodahl** [comment]
I'd like one more mentor to review, but I will make sure to merge before the end of GSSoC.

**Comment by Rukmini-Meda** [comment]
@alodahl @Kundan28 My score has been reduced by 4 points. I checked yesterday morning, it was showing my right score, but when i checked today it shows 4 points less. I think there has been a mistake since this PR was merged on June 1st. My score is not being considered for this PR. Can you please talk to the girlscript team and resolve to add these 4 points?

**Comment by Kundan28** [comment]
> @alodahl @Kundan28 My score has been reduced by 4 points. I checked yesterday morning, it was showing my right score, but when i checked today it shows 4 points less. I think there has been a mistake since this PR was merged on June 1st. My score is not being considered for this PR. Can you please talk to the girlscript team and resolve to add these 4 points?

Your PR was merged after the end date of GSSoC20, that's why it's showing you 4 points less.

**Comment by Rukmini-Meda** [comment]
@Kundan28 But I was told by @alodahl that it will be made sure that the PR be merged before GSSoC. Is it possible to discuss about the same once?

**Comment by Kundan28** [comment]
> @Kundan28 But I was told by @alodahl that it will be made sure that the PR be merged before GSSoC. Is it possible to discuss about the same once?

But unfortunately the PR was merged after the GSSoC was ended so we have to talk with the team regarding that, and we have already pinged them but didn't got any reply from their side.

**Comment by Kundan28** [comment]
> @Kundan28 But I was told by @alodahl that it will be made sure that the PR be merged before GSSoC. Is it possible to discuss about the same once?

But unfortunately the PR was merged after the GSSoC was ended so we have to talk with the team regarding that, and we have already pinged them but didn't got any reply from their side.
